README.md
# Tinkerforge Nagios Checks

Here you can find a collection of self written nagios checks for the tinkerforge system.
As soon as i have the possibility to get a new brick or bricklet, i will update this repository
with new checks. Feel free to send me some!

## distance-ir-bricklet

The Nagios Tinkerforge Distance IR Bricklet check was built to get the distance the bricklet measures
and compares that value against the minimum desired distance and the maximum desired distance given by
a single desired distance and a threshold. Its provided as the `check_tf_distance.sh` script

The threshold can be calculated for each distance-ir-bricklet itself using the provided `tf_distance_ir_bricklet_stddev_calc.sh`
script. It measures the distance once a second for a user-defined amount of total measurements and calculates the first standard
deviation from it which will then provide the threshold for the nagios check.

For a good standard deviation, built up your bricklet with the desired distance it should measure later and run the script with some thousands of measurements (i recommend 7200 whereas the script will run about 2 hours in that case).

The nagios check itself then uses the standard deviation provided as the threshold to generate warnings if the measurement is larger than the first standard deviation and a critical if its more than the second standard deviation or twice the threshold.

### How to get the standard deviation/check threshold?

1. Built up your test case with the desired distance you want to measure
2. Run the tf_distance_ir_bricklet_stddev_calc.sh script to get measurements and to calculate the standard deviation
    - `bash tf_distance_ir_bricklet_stddev_calc.sh <distance_ir_bricklet_UID> <sample_points>`
    - `bash tf_distance_ir_bricklet_stddev_calc.sh hVu 7200`
3. The script will output the standard deviation which is the threshold for the check itself
4. There is an example file (distance_ir_samples_20131106_232249.dat) with 30.000 sample points from a GP2Y0A02 sensor and a standard deviation of about 5

### How to use the check?

1. Obviously you should know how to use nagios checks and this one works the same. Just an example.
    - `check_tf_distance.sh <bricklet_uid> <desired_distance_in_mm> <measurement_threshold_in_mm>`
    - in action: `check_tf_distance.sh hVu 1000 12`

### How does a Nagios message look like?

#### Critical State

```
--SERVICE-ALERT-------------------
-
- Hostaddress: 10.0.0.1
- Hostname:    monitor.example.com
- Service:     Check Rack Door Distance: monitor.example.com
- - - - - - - - - - - - - - - - -
- State:       CRITICAL
- Date:        2013-11-06 18:30:14
- Output:      CRITICAL - Measured: 1438 mm, Expected: 1010 mm, Threshold: 5 mm
-
----------------------------------
```

#### Normal State

```
--SERVICE-ALERT-------------------
-
- Hostaddress: 10.0.0.1
- Hostname:    monitor.example.com
- Service:     Check Rack Door Distance: monitor.example.com
- - - - - - - - - - - - - - - - -
- State:       OK
- Date:        2013-11-06 18:34:50
- Output:      OK - Measured: 1010 mm, Expected: 1010 mm, Threshold: 5 mm
-
----------------------------------
```

## Copyright and License

(c) 2013 Martin Seener

Released under the GNU GPLv2


CHANGELOG.md
# Tinkerforge Nagios Checks

* **2013-11-07**: Added an example sample points file generated from my GP2Y0A02 Sensor with a real distance of 1010mm and with 30.000 sample points measured in about 9 hours with a standard deviation of ~5 (4.616405597) (distance_ir_samples_20131106_232249.dat)

* **2013-11-06**: Added small examples for a critical and normal check state

* **2013-11-05**: Added initial version of *nagios_tf_distance-ir-bricklet check* and *distance-ir-bricklet standard deviation calculator*

distance-ir-bricklet/tf_distance_ir_bricklet_stddev_calc.sh
#!/usr/bin/env bash

# tf-distance-ir-bricklet-stddev-calc.sh
# v1.0 (c) 2013, Martin Seener (martin@seener.de)

# Calculates the standard deviation of the distance-ir-bricklet
# by measuring a user-defined time in seconds whereas every second a sample is recorded
# to a sample_file for later use.

PROGNAME=`basename $0`
VERSION="Version 1.0,"
AUTHOR="2013, Martin Seener (martin@seener.de)"

print_version() {
  echo "$VERSION $AUTHOR"
}

print_help() {
  print_version $PROGNAME $VERSION
  echo ""
  echo "$PROGNAME is a tool to calculate the standard deviation of tinkerforge's distance-ir-bricklet to be used for the nagios distance-check as the threshold."
  echo "This tool requires the Brick Daemon 'brickd' to be installed as well as the tinkerforge api shell bindings."
  echo "You can find them here: http://www.tinkerforge.com/de/doc/index.html#software"
  echo ""
  echo "Usage: $0 <bricklet_uid> <desired_sample_points> | $0 hDu 120"
  echo "The example would measure 120 sample points (about one each second) and calculate the standard deviation out of it."
  echo ""
}

case "$1" in
  --help|-h)
    print_help
    exit $STATE_UNKNOWN
    ;;
  --version|-v)
    print_version $PROGNAME $VERSION
    exit $STATE_UNKNOWN
    ;;
  *)
    ;;
esac

### FUNCTIONS

check_prerequisites() {
  TINKERFORGE=$(which tinkerforge)
  if [ "$?" -ne 0 ]; then
    echo "Unable to find the tinkerforge shell bindings. Please install them first and dont miss brickd."
    echo ""
    print_help
    exit 3
  fi
  $TINKERFORGE enumerate | grep $1 > /dev/null 2>&1
  if [ "$?" -ne 0 ]; then
    echo "Unable to find the distance-ir-bricklet with the UID: $1"
    echo ""
    print_help
    exit 3
  fi
}

measure_samples() {
  touch $3 > /dev/null 2>&1
  if [ "$?" -ne 0 ]; then
    echo "Unable to create the local sample file: ./$SAMPLEFILE. Maybe i dont have write permissions?"
    echo ""
    print_help
    exit 3
  fi
  for ((i=$2; i>0; i--))
  do
    $TINKERFORGE call distance-ir-bricklet $1 get-distance | cut -d'=' -f2 >> $3
    sleep 1
  done
}

arithmetic_mean () {
  local RT=0         # Running total
  local AM=0         # Arithmetic mean
  local SCT=0        # Sample Point count
  while read VALUE
  do
    RT=$(echo "scale=$SCALE; $RT + $VALUE" | bc)
    (( SCT++ ))
  done < $1
  AM=$(echo "scale=$SCALE; $RT / $SCT" | bc)
  echo "$AM#$SCT"
}

standard_deviation() {
  ARITHMEAN=$(arithmetic_mean $1)

  MEAN1=$(echo $ARITHMEAN | cut -d'#' -f1)  # Arithmetic mean
  MEAN2=$(echo $ARITHMEAN | cut -d'#' -f2)  # Sample Point count
  SUM2=0                                    # Sum of squared differences ("variance").
  AVG2=0                                    # Average of $sum2.
  STDDEV=0                                  # Standard Deviation.

  while read VALUE
  do
    DIFF=$(echo "scale=$SCALE; $MEAN1 - $VALUE" | bc)
    DIFF2=$(echo "scale=$SCALE; $DIFF * $DIFF" | bc)
    SUM2=$(echo "scale=$SCALE; $SUM2 + $DIFF2" | bc)
  done < $1

  AVG2=$(echo "scale=$SCALE; $SUM2 / $MEAN2" | bc)
  STDDEV=$(echo "scale=$SCALE; sqrt($AVG2)" | bc)
  echo $STDDEV
}

initialize() {
  if [ -z "$1" ]; then
    echo "Sorry, but the first argument is missing, and has to be the UID of the distance-ir-bricklet!"
    echo ""
    print_help
    exit 3
  fi
  if [[ ! $2 =~ ^[0-9]+$ ]] || [ $2 -eq 0 ]; then
    echo "Sorry, but the second argument must be a positiver integer starting at 1 or higher!"
    echo ""
    print_help
    exit 3
  fi

  TIMESTAMP=$(date +%Y%m%d_%H%M%S)
  SAMPLEFILE="distance_ir_samples_$TIMESTAMP.dat"
  BRICKLETUID=$1
  SAMPLEPOINTS=$2
  SCALE=9      # Scale used by bc. Default 9 decimal places
}

### START

# Initialize with parameter-check and creation of the timestamp and sample file
initialize $1 $2
# Check if there is a distance-ir-bricklet there with the specified uid
check_prerequisites $BRICKLETUID
# Measure the samples
measure_samples $BRICKLETUID $SAMPLEPOINTS $SAMPLEFILE
# Calculate the standard deviation
echo "The standard deviation for the sample points in: $SAMPLEFILE is $(standard_deviation $SAMPLEFILE)"

# We are done
exit 0


distance-ir-bricklet/check_tf_distance.sh
#!/usr/bin/env bash

# check_tf_distance.sh v1.0
# (c) 2013 - Martin Seener (martin@seener.de)

# This Nagios check will connect to your tinkerforge distance-ir-bricklet, read the measured distance
# and alert you when the distance is more or less than a given distance using a threshold.
# The threshold can be calculated using tf_distance_ir_bricklet_stddev_calc.sh script. For that built up
# your bricklet with the distance you want to measure and run that script for one or two hours (3600 or 7200)
# measurements. It will calculate the standard deviation for you which then can be used as the threshold.

# This check can be used for example for a distance check, if a door in a server rack was opened, therefore
# the distance between the bricklet and the door has de- or increased.

PROGNAME=$(basename $0)
VERSION="Version 1.0,"
AUTHOR="2013, Martin Seener (martin@seener.de)"

# Nagios States
STATE_OK=0
STATE_WARN=1
STATE_CRIT=2
STATE_UNKNOWN=3

print_version() {
  echo "$VERSION $AUTHOR"
}

print_help() {
  print_version $PROGNAME $VERSION
  echo ""
  echo "$PROGNAME is a Nagios plugin to check the distance measured by a tinkerforge distance-ir-bricklet."
  echo "This check requires the Brick Daemon 'brickd' to be installed as well as the tinkerforge api shell bindings."
  echo "You can find them here: http://www.tinkerforge.com/de/doc/index.html#software"
  echo ""
  echo "Usage: $0 <bricklet_uid> <desired_distance_in_mm> <measurement_threshold_in_mm> | $0 hDu 1000 5"
  echo ""
}

check_prerequisites() {
  TINKERFORGE=$(which tinkerforge)
  if [ "$?" -ne 0 ]; then
    echo "Unable to find the tinkerforge shell bindings. Please install them first and dont miss brickd."
    echo ""
    print_help
    exit $STATE_UNKNOWN
  fi
  $TINKERFORGE enumerate | grep $1 > /dev/null 2>&1
  if [ "$?" -ne 0 ]; then
    echo "Unable to find the distance-ir-bricklet with the UID: $1"
    echo ""
    print_help
    exit $STATE_UNKNOWN
  fi
}

initialize() {
  if [ -z "$1" ]; then
    echo "Sorry, but the first argument is missing, and has to be the UID of the distance-ir-bricklet!"
    echo ""
    print_help
    exit $STATE_UNKNOWN
  fi
  if [[ ! $2 =~ ^[0-9]+$ ]] || [ $2 -eq 0 ]; then
    echo "Sorry, but the second argument must be a positiver integer starting at 1 or higher!"
    echo ""
    print_help
    exit $STATE_UNKNOWN
  fi
  if [[ ! $3 =~ ^[0-9]+$ ]] || [ $2 -eq 0 ]; then
    echo "Sorry, but the second argument must be a positiver integer starting at 1 or higher!"
    echo ""
    print_help
    exit $STATE_UNKNOWN
  fi

  TIMESTAMP=$(date +%Y%m%d_%H%M%S)
}

measure_distance() {
  DISTANCE=$($TINKERFORGE call distance-ir-bricklet $1 get-distance | cut -d'=' -f2)

  # Calculate minimum distance
  MIN_DESIRED=$(echo "scale=0; ($2 - $3)/1" | bc)
  MAX_DESIRED=$(echo "scale=0; ($2 + $3)/1" | bc)
  if [[ $DISTANCE -ge $MIN_DESIRED ]] && [[ $DISTANCE -le $MAX_DESIRED ]]; then
    echo "OK - Measured: $DISTANCE mm, Expected: $2 mm, Threshold: $3 mm|measured=$DISTANCE,expected=$2,threshold=$3"
    exit $STATE_OK
  fi

  # Measured distance is not in the range, so we check the doubled standard deviation
  MIN_DESIRED2=$(echo "scale=0; ($2 - (2 * $3))/1" | bc)
  MAX_DESIRED2=$(echo "scale=0; ($2 + (2 * $3))/1" | bc)
  if [[ $DISTANCE -ge $MIN_DESIRED2 ]] && [[ $DISTANCE -le $MAX_DESIRED2 ]]; then
    echo "WARNING - Measured: $DISTANCE mm, Expected: $2 mm, Threshold: $3 mm|measured=$DISTANCE,expected=$2,threshold=$3"
    exit $STATE_WARN
  else
    echo "CRITICAL - Measured: $DISTANCE mm, Expected: $2 mm, Threshold: $3 mm|measured=$DISTANCE,expected=$2,threshold=$3"
    exit $STATE_CRIT
  fi
}

case "$1" in
  --help|-h)
    print_help
    exit $STATE_UNKNOWN
    ;;
  --version|-v)
    print_version $PROGNAME $VERSION
    exit $STATE_UNKNOWN
    ;;
  *)
    ;;
esac

# Initialize the Check
initialize $1 $2 $3
# Check prerequisites
check_prerequisites $1
# Measure the distance
measure_distance $1 $2 $3
